From f968a6900cf72c197326ccc06d8f3356c17fbb0d Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Wed, 22 May 2019 15:05:55 +0800
Subject: [PATCH] DRVLIB_SW-2340 - Enlarge gphy memory size to 32MB for
 workaround

This is to workaround gphy memory corruption issue.

We reserve full 16MB for gphy (additional 16MB is needed for alignment),
to avoid other place using the same memory base as gphy.

This patch needs to be reverted once we have fixed in
firmware side.
---
 drivers/net/ethernet/lantiq/xrx500_phy_fw.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
index 9d8f172d6b50..43ae44d40ffa 100644
--- a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
+++ b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
@@ -54,6 +54,7 @@ struct xway_gphy_data {
 		int (*boot_func)(struct xway_gphy_data *);
 		int (*dt_parse_func)(struct xway_gphy_data *);
 		int align;
+		int fixed_mem_size;
 	} *soc_data;
 };
 
@@ -487,11 +488,17 @@ static int xway_gphy_load(struct xway_gphy_data *priv)
 		return -EIO;
 	}
 
+	if (priv->soc_data->fixed_mem_size)
+		size = priv->soc_data->fixed_mem_size;
+	else
+		size = fw->size;
+
 	/**
 	 * GPHY cores need the firmware code in a persistent and contiguous
 	 * memory area with a boundary aligned start address.
 	 */
-	size = fw->size + priv->soc_data->align;
+	size += priv->soc_data->align;
+
 	virt_addr = dmam_alloc_coherent(priv->dev, size, &priv->dma_addr,
 					GFP_KERNEL);
 	if (!virt_addr) {
@@ -562,7 +569,12 @@ static struct xway_gphy_soc_data xrx500_gphy_data = {
 static struct xway_gphy_soc_data prx300_gphy_data = {
 	.boot_func = &prx300_gphy_boot,
 	.dt_parse_func = &prx300_dt_parse,
-	.align = 128 * 1024,
+
+	/* To workaround GPHY memory corruption issue,
+	 * we allocate 16MB (plus additional 16MB for alignment)
+	 */
+	.align = 16 * 1024 * 1024,
+	.fixed_mem_size = 16 * 1024 * 1024,
 };
 
 static const struct of_device_id xway_phy_match[] = {
